### PR TITLE
Fix slice bytestring and add builtin tests

### PR DESCRIPTION
**Pull request by ak3n**
@bezirg noticed yesterday that our `sliceByteString` differs from slice operations in other languages:

```
python: "hello world"[0:1] == java's "hello world".substring(0,1) == javascript  "hello world".slice(0,1) == "h"
```

While in Plutus we do `"hello world".slice(0,1) == "he"` which is incorrect.

- This PR changes the behaviour to the same as in `slice` from `vector` package.
- Add tests.

---

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

**Comment by kwxm** [comment]
It's debatable whether this behaviour is more common.  What's happening above is that looking at the substring from `a` to `b` is giving you the elements of index `i` for `a <= i < b` (and I think this is probably because arrays start at 0 so people are used to the `(for i=0; i<n; i++)` idiom from C). If you look at OCaml,R, and awk, they all have a subarray operation that gives you the elements for `a <= i <= b`.  In C++ and bash you give it a start position and a length.

I like it the way it is, but I don't know if it's a big deal.

File: plutus-tx/src/PlutusTx/Builtins.hs [hunk 1]
```diff
@@ -88,19 +88,19 @@ consByteString :: Integer -> BuiltinByteString -> BuiltinByteString
 consByteString n bs = BI.consByteString (toBuiltin n) bs
 
 {-# INLINABLE sliceByteString #-}
--- | Returns the substring of a 'ByteString'.
+-- | Returns the substring of a 'ByteString' from index 'from' to index 'to'.
```
Comment (kwxm) [inline plutus-tx/src/PlutusTx/Builtins.hs:91]:
```suggestion
-- | Returns the substring of a 'ByteString' from index 'from' up to but not including index 'to'.
```

**Comment by michaelpj** [comment]
Hmm, I wonder if the more usual Haskell idiom is start index and length (which tbh avoids these issues indeed...). e.g. https://hackage.haskell.org/package/vector-0.12.3.0/docs/Data-Vector.html#v:slice

**Comment by ak3n** [comment]
@michaelpj sounds like the best option. I have changed the behaviour to the same as in vector package.

File: plutus-tx/src/PlutusTx/Builtins.hs [hunk 2]
```diff
@@ -88,19 +88,19 @@ consByteString :: Integer -> BuiltinByteString -> BuiltinByteString
...
 -- | Returns the n length prefix of a 'ByteString'.
 takeByteString :: Integer -> BuiltinByteString -> BuiltinByteString
-takeByteString n bs = BI.sliceByteString 0 (toBuiltin n - 1) bs
+takeByteString n bs = BI.sliceByteString 0 (toBuiltin n) bs
 
 {-# INLINABLE dropByteString #-}
 -- | Returns the suffix of a 'ByteString' after n elements.
 dropByteString :: Integer -> BuiltinByteString -> BuiltinByteString
-dropByteString n bs = BI.sliceByteString (toBuiltin n) (BI.lengthOfByteString bs - 1) bs
+dropByteString n bs = BI.sliceByteString (toBuiltin n) (BI.lengthOfByteString bs) bs
```
Comment (michaelpj) [inline plutus-tx/src/PlutusTx/Builtins.hs:103]: should be from `n` to `length bs - n`, right?

**Review by michaelpj — COMMENTED** [review]
The definitions of take and drop are annoyingly subtle. Maybe add a test or two for them in `plutus-tx/test`? There are some tests for the stuff defined in `plutus-tx` there.

**Comment by michaelpj** [comment]
I'm happy with the definition change, and it doesn't make the costing any more complex (the "length" argument is effectively an index, with one more addition in there).

File: plutus-tx/src/PlutusTx/Builtins.hs (hunk 2 above)
Comment (ak3n) [reply plutus-tx/src/PlutusTx/Builtins.hs:103]: Oops, right! I checked it on a short example and it looked fine. Thanks!

File: plutus-tx/src/PlutusTx/Builtins.hs [hunk 3]
```diff
@@ -88,19 +88,19 @@ consByteString :: Integer -> BuiltinByteString -> BuiltinByteString
...
 -- | Returns the n length prefix of a 'ByteString'.
 takeByteString :: Integer -> BuiltinByteString -> BuiltinByteString
-takeByteString n bs = BI.sliceByteString 0 (toBuiltin n - 1) bs
+takeByteString n bs = BI.sliceByteString 0 (toBuiltin n) bs
 
 {-# INLINABLE dropByteString #-}
 -- | Returns the suffix of a 'ByteString' after n elements.
 dropByteString :: Integer -> BuiltinByteString -> BuiltinByteString
-dropByteString n bs = BI.sliceByteString (toBuiltin n) (BI.lengthOfByteString bs - 1) bs
+dropByteString n bs = BI.sliceByteString (toBuiltin n) (BI.lengthOfByteString bs - n) bs
```
Comment (michaelpj) [inline plutus-tx/src/PlutusTx/Builtins.hs:103]: `minusInteger`

Comment (michaelpj) [reply plutus-tx/src/PlutusTx/Builtins.hs:103]: I think this will get the wrong `-` otherwise

Comment (michaelpj) [reply plutus-tx/src/PlutusTx/Builtins.hs:103]: Alternatively, maybe we should define these in `PlutusTx.Prelude`, where the right `-` will be in scope, and these aren't builtins any more anyway.

Comment (michaelpj) [reply plutus-tx/src/PlutusTx/Builtins.hs:103]: take and drop, that is

**Review by michaelpj — APPROVED** [review]

File: plutus-tx/src/PlutusTx/Builtins.hs (hunk 3 above)
Comment (ak3n) [reply plutus-tx/src/PlutusTx/Builtins.hs]: Good catch!

It's my second time of using the wrong operation. Maybe it's worth to add `NoImplicitPrelude` everywhere and use a special module `AllowedPrelude`? 🤔

**Pull request merged**